emacs/config.org
#+TITLE: Emacs Configuration

* Package Management

** Elpaca

#+begin_src emacs-lisp
(defvar elpaca-installer-version 0.12)
(defvar elpaca-directory (expand-file-name "elpaca/" user-emacs-directory))
(defvar elpaca-builds-directory (expand-file-name "builds/" elpaca-directory))
(defvar elpaca-sources-directory (expand-file-name "sources/" elpaca-directory))
(defvar elpaca-order '(elpaca :repo "https://github.com/progfolio/elpaca.git"
                              :ref nil :depth 1 :inherit ignore
                              :files (:defaults "elpaca-test.el" (:exclude "extensions"))
                              :build (:not elpaca-activate)))
(let* ((repo  (expand-file-name "elpaca/" elpaca-sources-directory))
       (build (expand-file-name "elpaca/" elpaca-builds-directory))
       (order (cdr elpaca-order))
       (default-directory repo))
  (add-to-list 'load-path (if (file-exists-p build) build repo))
  (unless (file-exists-p repo)
    (make-directory repo t)
    (when (<= emacs-major-version 28) (require 'subr-x))
    (condition-case-unless-debug err
        (if-let* ((buffer (pop-to-buffer-same-window "*elpaca-bootstrap*"))
                  ((zerop (apply #'call-process `("git" nil ,buffer t "clone"
                                                  ,@(when-let* ((depth (plist-get order :depth)))
                                                      (list (format "--depth=%d" depth) "--no-single-branch"))
                                                  ,(plist-get order :repo) ,repo))))
                  ((zerop (call-process "git" nil buffer t "checkout"
                                        (or (plist-get order :ref) "--"))))
                  (emacs (concat invocation-directory invocation-name))
                  ((zerop (call-process emacs nil buffer nil "-Q" "-L" "." "--batch"
                                        "--eval" "(byte-recompile-directory \".\" 0 'force)")))
                  ((require 'elpaca))
                  ((elpaca-generate-autoloads "elpaca" repo)))
            (progn (message "%s" (buffer-string)) (kill-buffer buffer))
          (error "%s" (with-current-buffer buffer (buffer-string))))
      ((error) (warn "%s" err) (delete-directory repo 'recursive))))
  (unless (require 'elpaca-autoloads nil t)
    (require 'elpaca)
    (elpaca-generate-autoloads "elpaca" repo)
    (let ((load-source-file-function nil)) (load "./elpaca-autoloads"))))
(add-hook 'after-init-hook #'elpaca-process-queues)
(elpaca `(,@elpaca-order))
#+end_src

** Use-package Support

#+begin_src emacs-lisp
;; Install use-package support
(elpaca elpaca-use-package
  ;; Enable use-package :ensure support for Elpaca.
  (elpaca-use-package-mode))
#+end_src


** How to Use Elpaca
#+begin_src emacs-lisp
;;When installing a package used in the init file itself,
;;e.g. a package which adds a use-package key word,
;;use the :wait recipe keyword to block until that package is installed/configured.
;;For example:
;;(use-package general :ensure (:wait t) :demand t)

;; Expands to: (elpaca evil (use-package evil :demand t))
;;(use-package evil :ensure t :demand t)

;;Turns off elpaca-use-package-mode current declaration
;;Note this will cause evaluate the declaration immediately. It is not deferred.
;;Useful for configuring built-in emacs features.
;;(use-package emacs :ensure nil :config (setq ring-bell-function #'ignore))
#+end_src



* Editing

** Evil

#+begin_src emacs-lisp
  (use-package evil
    :ensure t
    :init
    (setq evil-want-integration t)  ;; for evil-collection
    (setq evil-want-keybinding nil) ;; for evil-collection
    :config
    (evil-mode))
#+end_src

*** Evil Collection

#+begin_src emacs-lisp
  (use-package evil-collection
    :after evil
    :ensure t
    :config
    (evil-collection-init))
#+end_src

** Python Packages

*** Elpy

Install ~elpy~ - the Emacs Python Development Environment

#+begin_src emacs-lisp
  (use-package elpy
    :ensure t
    :init
    (elpy-enable))
#+end_src

** Keybinds

*** General

General makes it easier to set keybindings

#+begin_src emacs-lisp
  (use-package general
    :ensure t
    :config
    (general-evil-setup)

    ;; set up 'SPC' as the global leader key
    (general-create-definer char/leader-keys
      :states '(normal insert visual emacs)
      :keymaps 'override
      :prefix "SPC" ;; set leader
      :global-prefix "M-SPC") ;; access leader in insert mode

    (char/leader-keys
      "b" '(:ignore t :wk "buffer")
      "bb" '(switch-to-buffer :wk "Switch buffer")
      "bk" '(kill-this-buffer :wk "Kill this buffer")
      "bn" '(next-buffer :wk "Next buffer")
      "bp" '(previous-buffer :wk "Previous buffer")
      "br" '(revert-buffer :wk "Reload buffer"))

    (char/leader-keys
      "f" '(:ignore t :wk "file")
      "ff" '(find-file :wk "Find file"))

    (char/leader-keys
      "h" '(:ignore t :wk "help")
      "hh" '(help-for-help :wk "Help for help")
      "hf" '(describe-function :wk "Describe function")
      "hx" '(describe-function :wk "Describe command")
      "hv" '(describe-variable :wk "Describe variable")
      "hk" '(describe-key :wk "Describe key")
      "hc" '(describe-key-briefly :wk "Describe key briefly"))

  )
#+end_src

* GUI

** Theme

Catppuccin theme because its cute

#+begin_src emacs-lisp
  (use-package catppuccin-theme
    :ensure t
    :config
    (load-theme 'catppuccin :no-confirm))
#+end_src

** Line Numbers

Relative line numbers, duh

#+begin_src emacs-lisp
  (setq display-line-numbers 'relative)
  (global-display-line-numbers-mode t)
#+end_src

* File Management

** Backups

Disable backup files - since anything important I track with version control systems (i.e. git)

#+begin_src emacs-lisp
(setq make-backup-files 0)
#+end_src

* Help

** Which Key

#+begin_src emacs-lisp
    (setq which-key-allow-evil-operators t ;; Enable popups to show for evil operators
          which-key-show-early-on-C-h t ;; Trigger C-h popup before its done automatically
          which-key-side-window-location 'bottom
          which-key-sort-order #'which-key-key-order-alpha
          which-key-sort-uppercase-first nil
          which-key-add-column-padding 1
          which-key-max-display-columns nil
          which-key-min-display-lines 6
          which-key-side-window-slot -10
          which-key-side-window-max-height 0.25
          which-key-idle-delay 0.8
          which-key-max-description-length 25
          which-key-allow-imprecise-window-fit t
          which-key-separator " → " 
  	which-key-idle-secondary-delay 0.05)
    (which-key-mode)
#+end_src


emacs/config.el
(defvar elpaca-installer-version 0.12)
(defvar elpaca-directory (expand-file-name "elpaca/" user-emacs-directory))
(defvar elpaca-builds-directory (expand-file-name "builds/" elpaca-directory))
(defvar elpaca-sources-directory (expand-file-name "sources/" elpaca-directory))
(defvar elpaca-order '(elpaca :repo "https://github.com/progfolio/elpaca.git"
                              :ref nil :depth 1 :inherit ignore
                              :files (:defaults "elpaca-test.el" (:exclude "extensions"))
                              :build (:not elpaca-activate)))
(let* ((repo  (expand-file-name "elpaca/" elpaca-sources-directory))
       (build (expand-file-name "elpaca/" elpaca-builds-directory))
       (order (cdr elpaca-order))
       (default-directory repo))
  (add-to-list 'load-path (if (file-exists-p build) build repo))
  (unless (file-exists-p repo)
    (make-directory repo t)
    (when (<= emacs-major-version 28) (require 'subr-x))
    (condition-case-unless-debug err
        (if-let* ((buffer (pop-to-buffer-same-window "*elpaca-bootstrap*"))
                  ((zerop (apply #'call-process `("git" nil ,buffer t "clone"
                                                  ,@(when-let* ((depth (plist-get order :depth)))
                                                      (list (format "--depth=%d" depth) "--no-single-branch"))
                                                  ,(plist-get order :repo) ,repo))))
                  ((zerop (call-process "git" nil buffer t "checkout"
                                        (or (plist-get order :ref) "--"))))
                  (emacs (concat invocation-directory invocation-name))
                  ((zerop (call-process emacs nil buffer nil "-Q" "-L" "." "--batch"
                                        "--eval" "(byte-recompile-directory \".\" 0 'force)")))
                  ((require 'elpaca))
                  ((elpaca-generate-autoloads "elpaca" repo)))
            (progn (message "%s" (buffer-string)) (kill-buffer buffer))
          (error "%s" (with-current-buffer buffer (buffer-string))))
      ((error) (warn "%s" err) (delete-directory repo 'recursive))))
  (unless (require 'elpaca-autoloads nil t)
    (require 'elpaca)
    (elpaca-generate-autoloads "elpaca" repo)
    (let ((load-source-file-function nil)) (load "./elpaca-autoloads"))))
(add-hook 'after-init-hook #'elpaca-process-queues)
(elpaca `(,@elpaca-order))

;; Install use-package support
(elpaca elpaca-use-package
  (elpaca-use-package-mode))

;; Optional: faster parallel installs
(setq elpaca-queue-limit 30)

(use-package evil
  :ensure t
  :defer nil
  :init
  (setq evil-want-integration t
        evil-want-keybinding nil
        evil-vsplit-window-right t
        evil-split-window-below t)
  :config
  (evil-mode))

(use-package evil-collection
  :ensure t
  :after evil
  :config
  (setq evil-collection-mode-list '(dashboard dired ibuffer))
  (evil-collection-init))

(defvar char/emacs-config-org
  (expand-file-name "config.org" user-emacs-directory))

(use-package general
  :ensure t
  :after evil
  :demand t
  :config
  (general-evil-setup))
(elpaca-wait)

  ;; set up 'SPC' as the global leader key
  (general-create-definer char/leader-keys
    :states '(normal visual)
    :keymaps 'override
    :prefix "SPC"
    :global-prefix "M-SPC")

    ;; Buffer keys
  (defun char/open-scratch-buffer ()
    (interactive)
    (switch-to-buffer (get-buffer-create "*scratch*")))
  (char/leader-keys
    "b" '(:ignore t :wk "buffer")
    "bb" '(switch-to-buffer :wk "Switch buffer")
    "bk" '(kill-current-buffer :wk "Kill current buffer")
    "bn" '(next-buffer :wk "Next buffer")
    "bp" '(previous-buffer :wk "Previous buffer")
    "br" '(revert-buffer :wk "Reload buffer")
    "bi" '(ibuffer :wk "Ibuffer")
    "bs" '(char/open-scratch-buffer :wk "Open Scratch"))

    ;; File keys
  (defun char/find-org-config ()
    (interactive)
    (find-file char/emacs-config-org))
  (defun char/reload-config ()
    (interactive)
    (org-babel-tangle-file char/emacs-config-org)
    (load-file (expand-file-name "init.el" user-emacs-directory)))
  (char/leader-keys
    "f" '(:ignore t :wk "File")
    "ff" '(find-file :wk "Find File")
    "fo" '(:ignore t :wk "Open File")
    "foc" '(char/find-org-config :wk "Org Config")
    "fr" '(char/reload-config :wk "Reload Config"))

    ;; Help keys
  (char/leader-keys
    "h" '(:ignore t :wk "Help")
    "hk" '(describe-key :wk "Describe Key")
    "hc" '(describe-key-briefly :wk "Describe Key Briefly")
    "hx" '(describe-command :wk "Describe Command")
    "hf" '(describe-function :wk "Describe Function")
    "hv" '(describe-variable :wk "Describe Variable")
    "h?" '(help-for-help :wk "Help For Help"))

  ;; Eval keys
  (char/leader-keys
  "e" '(:ignore t :wk "Evaluate")
  "eb" '(eval-buffer :wk "Evaluate Elisp in Buffer")
  "ed" '(eval-defun :wk "Evaluate Defun Containing or After Point")
  "ee" '(eval-expression :wk "Evaluate an Elisp Expression")
  "el" '(eval-last-sexp :wk "Evaluate Elisp Expression Before Point")
  "er" '(eval-region :wk "Evaluate Elisp in Region"))

;; Editing keys
(char/leader-keys
  "i" '(:ignore t :wk "Edit")
  "ic" '(comment-line :wk "Toggle comment"))

(global-set-key (kbd "C-=") 'text-scale-increase)
(global-set-key (kbd "C--") 'text-scale-decrease)

(set-face-attribute 'default nil
		    :font "JetBrains Mono"
		    :height 110
		    :weight 'medium)
(set-face-attribute 'variable-pitch nil
		    :font "JetBrains Mono"
		    :height 120
		    :weight 'medium)
(set-face-attribute 'fixed-pitch nil
		    :font "JetBrains Mono"
		    :height 110
		    :weight 'medium)
;; Makes commented text italicized (is that a real word?).
;; Your font must have an italic face available.
(set-face-attribute 'font-lock-comment-face nil
		    :slant 'italic)

;; This sets the default font on all graphical frames created after restarting Emacs.
;; Does the same thing as 'set-face-attribute default' above, but emacsclient fonts
;; are not right unless I also add this method of setting the default font.
(add-to-list 'default-frame-alist '(font . "JetBrains Mono-11"))

;; Uncomment the following line if line spacing needs adjusting.
(setq-default line-spacing 0.12)

(menu-bar-mode 0)
(tool-bar-mode 0)
(scroll-bar-mode 0)

(global-display-line-numbers-mode 1)
(setq display-line-numbers-type 'relative)
(global-visual-line-mode t)

(use-package which-key
  :init
    (which-key-mode 1)
  :config
  (setq which-key-side-window-location 'bottom
	  which-key-sort-order #'which-key-key-order-alpha
	  which-key-sort-uppercase-first nil
	  which-key-add-column-padding 1
	  which-key-max-display-columns nil
	  which-key-min-display-lines 6
	  which-key-side-window-slot -10
	  which-key-side-window-max-height 0.25
	  which-key-idle-delay 0.8
	  which-key-max-description-length 45
	  which-key-allow-imprecise-window-fit t
	  which-key-separator " → " ))

(use-package toc-org
  :ensure t
  :hook (org-mode . toc-org-enable))

(use-package org-bullets
  :ensure t
  :hook (org-mode . org-bullets-mode))

(electric-indent-mode 0)

(require 'org-tempo)
